plugins/java-decompiler/engine/src/org/jetbrains/java/decompiler/modules/decompiler/ExprProcessor.java

```diff
@@ -851,12 +851,12 @@ public static boolean getCastedExprent(Exprent exprent,
                                          TextBuffer buffer,
                                          int indent,
                                          boolean castNull,
-                                         boolean castAlways,
+                                         boolean requiresExactTypeMatch,
                                          BytecodeMappingTracer tracer) {
     VarType rightType = exprent.getExprType();
 
     boolean cast =
-      castAlways ||
+      (requiresExactTypeMatch && !leftType.equals(exprent.getExprType())) ||
```

Comment:
First, in many cases the expression is a part of a complex statement (e.g. multi-parameter method invocation). So whether or not the cast is needed should be decided on a call site.

Next, omitting a cast because of a type equality doesn't seem safe. Wouldn't it fail on an example from IDEA-135385?